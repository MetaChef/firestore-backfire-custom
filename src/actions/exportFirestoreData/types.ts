import { FirestoreConnectionOptions, FirestoreDataOptions } from "~/types";
import { StorageSourceOptions } from "~/services/StorageSourceService";

export type ExportFirestoreDataOptions = FirestoreConnectionOptions &
  FirestoreDataOptions &
  StorageSourceOptions;
