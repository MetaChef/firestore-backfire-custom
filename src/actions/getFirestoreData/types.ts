export interface GetFirestoreDataOptions {
  stringify?: boolean | number | undefined;
}
