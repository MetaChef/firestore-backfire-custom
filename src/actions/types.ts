export interface FirestoreDataOptions {
  paths?: string[];
  stringify?: number | boolean;
  count?: boolean;
  limit?: number;
}
