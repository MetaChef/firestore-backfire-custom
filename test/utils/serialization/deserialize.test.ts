import { Firestore } from "@google-cloud/firestore";

import { SerializedGeoPoint, SerializedQuery, SerializedTimestamp } from "~/types";

import { deserializeDocuments } from "~/utils/serialization/deserialize";

describe(deserializeDocuments.name, () => {
  let firestore: Firestore;

  beforeAll(() => {
    firestore = new Firestore({ projectId: "demo" });
  });

  it("deserializes plain documents", () => {
    const data = deserializeDocuments(
      firestore,
      JSON.stringify({
        path: "documents/1",
        data: {
          from: "example@email.com",
          to: ["recipient1@email.com", "recipient2@email.com"],
          content: "Dear John,\nHello world!",
          attachments: 2,
          tags: { work: true, important: true },
        },
      })
    );
    expect(data).toMatchSnapshot();
  });

  it("deserializes documents as arrays", () => {
    const data = deserializeDocuments(
      firestore,
      JSON.stringify([
        { path: "documents/1", data: {} },
        { path: "documents/2", data: {} },
        { path: "documents/3", data: {} },
      ])
    );
    expect(data).toMatchSnapshot();
  });

  it("deserializes documents with timestamps", () => {
    const seconds = new Date("2022-01-01").getTime() / 1000;
    const data = deserializeDocuments(
      firestore,
      JSON.stringify({
        path: "documents/1",
        data: {
          id: "1",
          date: <SerializedTimestamp>{ _seconds: seconds, _nanoseconds: 0 },
          nested: {
            dates: [
              <SerializedTimestamp>{ _seconds: seconds, _nanoseconds: 0 },
              <SerializedTimestamp>{ _seconds: seconds, _nanoseconds: 0 },
            ],
          },
        },
        timestamps: ["date", "nested.dates.0", "nested.dates.1"],
      })
    );
    expect(data).toMatchSnapshot();
    expect(data[0].data.date.toDate()).toMatchSnapshot();
  });

  it("deserializes documents with geopoints", () => {
    const data = deserializeDocuments(
      firestore,
      JSON.stringify({
        path: "documents/1",
        data: {
          id: "1",
          point: <SerializedGeoPoint>{ _latitude: 10, _longitude: 20 },
          nested: {
            places: [
              <SerializedGeoPoint>{ _latitude: 11, _longitude: 20 },
              <SerializedGeoPoint>{ _latitude: 12, _longitude: 20 },
            ],
          },
        },
        geopoints: ["point", "nested.places.0", "nested.places.1"],
      })
    );
    expect(data).toMatchSnapshot();
  });

  it("deserializes documents with document references", () => {
    const data = deserializeDocuments(
      firestore,
      JSON.stringify({
        path: "documents/1",
        data: {
          id: "1",
          reference: {
            _firestore: { projectId: "demo" },
            _path: { segments: ["documents", "2"] },
          },
          references: [
            {
              _firestore: { projectId: "demo" },
              _path: { segments: ["documents", "2"] },
            },
          ],
        },
        documents: ["reference", "references.0"],
      })
    );
    expect(data).toMatchSnapshot();
    const ref = firestore.doc("documents/2");
    expect(ref.isEqual(data[0].data.reference)).toBe(true);
    expect(ref.isEqual(data[0].data.references[0])).toBe(true);
  });

  it("deserializes documents with collection references", () => {
    const data = deserializeDocuments(
      firestore,
      JSON.stringify({
        path: "documents/1",
        data: {
          id: "1",
          reference: <SerializedQuery>{
            _firestore: { projectId: "demo" },
            _queryOptions: {
              parentPath: { segments: [] },
              collectionId: "documents",
              fieldFilters: [],
              fieldOrders: [],
            },
          },
          references: [
            <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [],
              },
            },
          ],
        },
        queries: ["reference", "references.0"],
      })
    );
    expect(data).toMatchSnapshot();
    const ref = firestore.collection("documents");
    expect(ref.isEqual(data[0].data.reference)).toBe(true);
    expect(ref.isEqual(data[0].data.references[0])).toBe(true);
  });

  describe("deserializes documents with queries", () => {
    it("field filter (where: comparison)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: ["documents", "1"] },
                collectionId: "threads",
                fieldFilters: [
                  {
                    op: "GREATER_THAN",
                    field: { segments: ["details", "count"] },
                    value: 10,
                  },
                  {
                    op: "LESS_THAN_OR_EQUAL",
                    field: { segments: ["details", "count"] },
                    value: 100,
                  },
                ],
                fieldOrders: [],
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents/1/threads");
      col.where("details.count", ">", 10).where("details.count", "<=", 100);

      // FIXME: Currently failing due to https://github.com/googleapis/nodejs-firestore/issues/1652
      // expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("field filter (where: equality)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: ["documents", "1"] },
                collectionId: "threads",
                fieldFilters: [
                  { field: { segments: ["id"] }, op: "EQUAL", value: "1" },
                  { field: { segments: ["user"] }, op: "NOT_EQUAL", value: "123" },
                ],
                fieldOrders: [],
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents/1/threads");
      col.where("id", "==", "1").where("user", "!=", "123");

      // FIXME: Currently failing due to https://github.com/googleapis/nodejs-firestore/issues/1652
      // expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("field filter (where: array)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: ["documents", "1"] },
                collectionId: "threads",
                fieldFilters: [
                  { field: { segments: ["id"] }, op: "IN", value: [15, 16, 17] },
                  {
                    field: { segments: ["count"] },
                    op: "ARRAY_CONTAINS",
                    value: [1, 2, 3, 4],
                  },
                ],
                fieldOrders: [],
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents/1/threads");
      col
        .where("id", "in", [15, 16, 17])
        .where("count", "array-contains", [1, 2, 3, 4]);

      // FIXME: Currently failing due to https://github.com/googleapis/nodejs-firestore/issues/1652
      // expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("field order (orderBy)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: ["documents", "1"] },
                collectionId: "threads",
                fieldFilters: [],
                fieldOrders: [
                  { field: { segments: ["id"] }, direction: "ASCENDING" },
                  { field: { segments: ["date"] }, direction: "DESCENDING" },
                ],
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents/1/threads");
      const query = col.orderBy("id", "asc").orderBy("date", "desc");
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("startAt (with integer proto field)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [
                  { field: { segments: ["id"] }, direction: "ASCENDING" },
                ],
                startAt: {
                  before: true,
                  values: [{ integerValue: 10 }],
                },
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col.orderBy("id", "asc").startAt(10);
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("startAfter (with nested map proto fields)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [
                  { field: { segments: ["properties"] }, direction: "DESCENDING" },
                ],
                startAt: {
                  before: false,
                  values: [
                    {
                      mapValue: {
                        fields: {
                          info: {
                            mapValue: {
                              fields: { strength: { stringValue: "high" } },
                            },
                          },
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col
        .orderBy("properties", "desc")
        .startAfter({ info: { strength: "high" } });
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("endAt (with boolean and array proto fields)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [
                  { field: { segments: ["properties"] }, direction: "ASCENDING" },
                ],
                endAt: {
                  before: false,
                  values: [
                    {
                      mapValue: {
                        fields: {
                          error: { booleanValue: true },
                          counts: {
                            arrayValue: {
                              values: [
                                { integerValue: 1 },
                                { integerValue: 2 },
                                { integerValue: 3 },
                              ],
                            },
                          },
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col
        .orderBy("properties")
        .endAt({ error: true, counts: [1, 2, 3] });
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("endBefore (with multiple proto fields)", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [
                  { field: { segments: ["average"] }, direction: "ASCENDING" },
                  { field: { segments: ["properties"] }, direction: "ASCENDING" },
                ],
                endAt: {
                  before: true,
                  values: [
                    { doubleValue: 12.5 },
                    {
                      mapValue: {
                        fields: {
                          color: { stringValue: "blue" },
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col
        .orderBy("average")
        .orderBy("properties")
        .endBefore(12.5, { color: "blue" });
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("limit", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [],
                limit: 10,
                limitType: 0,
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col.limit(10);
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("limitToLast", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [],
                limit: 10,
                limitType: 1,
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col.limitToLast(10);
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });

    it("offset", () => {
      const data = deserializeDocuments(
        firestore,
        JSON.stringify({
          path: "documents/1",
          data: {
            ref: <SerializedQuery>{
              _firestore: { projectId: "demo" },
              _queryOptions: {
                parentPath: { segments: [""] },
                collectionId: "documents",
                fieldFilters: [],
                fieldOrders: [],
                offset: 10,
              },
            },
          },
          queries: ["ref"],
        })
      );

      expect(data).toMatchSnapshot();

      const col = firestore.collection("documents");
      const query = col.offset(10);
      expect(query.isEqual(data[0].data.ref)).toBe(true);
    });
  });
});
