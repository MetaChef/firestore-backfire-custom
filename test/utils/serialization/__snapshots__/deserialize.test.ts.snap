// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserializeDocuments deserializes documents as arrays 1`] = `
Array [
  Object {
    "data": Object {},
    "path": "documents/1",
  },
  Object {
    "data": Object {},
    "path": "documents/2",
  },
  Object {
    "data": Object {},
    "path": "documents/3",
  },
]
`;

exports[`deserializeDocuments deserializes documents with collection references 1`] = `
Array [
  Object {
    "data": Object {
      "id": "1",
      "reference": CollectionReference {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
      "references": Array [
        CollectionReference {
          "_allowUndefined": false,
          "_firestore": Object {
            "projectId": "demo",
          },
          "_queryOptions": QueryOptions {
            "allDescendants": false,
            "collectionId": "documents",
            "converter": Object {
              "fromFirestore": [Function],
              "toFirestore": [Function],
            },
            "endAt": undefined,
            "fieldFilters": Array [],
            "fieldOrders": Array [],
            "kindless": false,
            "limit": undefined,
            "limitType": undefined,
            "offset": undefined,
            "parentPath": ResourcePath {
              "segments": Array [],
            },
            "projection": undefined,
            "requireConsistency": true,
            "startAt": undefined,
          },
          "_serializer": Serializer {
            "allowUndefined": false,
            "createInteger": [Function],
            "createReference": [Function],
          },
        },
      ],
    },
    "path": "documents/1",
    "queries": Array [
      "reference",
      "references.0",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with document references 1`] = `
Array [
  Object {
    "data": Object {
      "id": "1",
      "reference": DocumentReference {
        "_converter": Object {
          "fromFirestore": [Function],
          "toFirestore": [Function],
        },
        "_firestore": Object {
          "projectId": "demo",
        },
        "_path": ResourcePath {
          "segments": Array [
            "documents",
            "2",
          ],
        },
      },
      "references": Array [
        DocumentReference {
          "_converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "_firestore": Object {
            "projectId": "demo",
          },
          "_path": ResourcePath {
            "segments": Array [
              "documents",
              "2",
            ],
          },
        },
      ],
    },
    "documents": Array [
      "reference",
      "references.0",
    ],
    "path": "documents/1",
  },
]
`;

exports[`deserializeDocuments deserializes documents with geopoints 1`] = `
Array [
  Object {
    "data": Object {
      "id": "1",
      "nested": Object {
        "places": Array [
          GeoPoint {
            "_latitude": 11,
            "_longitude": 20,
          },
          GeoPoint {
            "_latitude": 12,
            "_longitude": 20,
          },
        ],
      },
      "point": GeoPoint {
        "_latitude": 10,
        "_longitude": 20,
      },
    },
    "geopoints": Array [
      "point",
      "nested.places.0",
      "nested.places.1",
    ],
    "path": "documents/1",
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries endAt (with boolean and array proto fields) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": Object {
            "before": false,
            "values": Array [
              Object {
                "mapValue": Object {
                  "fields": Object {
                    "counts": Object {
                      "arrayValue": Object {
                        "values": Array [
                          Object {
                            "integerValue": 1,
                          },
                          Object {
                            "integerValue": 2,
                          },
                          Object {
                            "integerValue": 3,
                          },
                        ],
                      },
                    },
                    "error": Object {
                      "booleanValue": true,
                    },
                  },
                },
              },
            ],
          },
          "fieldFilters": Array [],
          "fieldOrders": Array [
            FieldOrder {
              "direction": "ASCENDING",
              "field": FieldPath {
                "segments": Array [
                  "properties",
                ],
              },
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries endBefore (with multiple proto fields) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": Object {
            "before": true,
            "values": Array [
              Object {
                "doubleValue": 12.5,
              },
              Object {
                "mapValue": Object {
                  "fields": Object {
                    "color": Object {
                      "stringValue": "blue",
                    },
                  },
                },
              },
            ],
          },
          "fieldFilters": Array [],
          "fieldOrders": Array [
            FieldOrder {
              "direction": "ASCENDING",
              "field": FieldPath {
                "segments": Array [
                  "average",
                ],
              },
            },
            FieldOrder {
              "direction": "ASCENDING",
              "field": FieldPath {
                "segments": Array [
                  "properties",
                ],
              },
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries field filter (where: array) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "threads",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "id",
                ],
              },
              "op": "IN",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": Array [
                15,
                16,
                17,
              ],
            },
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "count",
                ],
              },
              "op": "ARRAY_CONTAINS",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": Array [
                1,
                2,
                3,
                4,
              ],
            },
          ],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [
              "documents",
              "1",
            ],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries field filter (where: comparison) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "threads",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "details",
                  "count",
                ],
              },
              "op": "GREATER_THAN",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": 10,
            },
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "details",
                  "count",
                ],
              },
              "op": "LESS_THAN_OR_EQUAL",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": 100,
            },
          ],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [
              "documents",
              "1",
            ],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries field filter (where: equality) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "threads",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "id",
                ],
              },
              "op": "EQUAL",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": "1",
            },
            FieldFilter {
              "field": FieldPath {
                "segments": Array [
                  "user",
                ],
              },
              "op": "NOT_EQUAL",
              "serializer": Serializer {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": "123",
            },
          ],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [
              "documents",
              "1",
            ],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries field order (orderBy) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "threads",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [
            FieldOrder {
              "direction": "ASCENDING",
              "field": FieldPath {
                "segments": Array [
                  "id",
                ],
              },
            },
            FieldOrder {
              "direction": "DESCENDING",
              "field": FieldPath {
                "segments": Array [
                  "date",
                ],
              },
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [
              "documents",
              "1",
            ],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries limit 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": 10,
          "limitType": 0,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries limitToLast 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": 10,
          "limitType": 1,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries offset 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": 10,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries startAfter (with nested map proto fields) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [
            FieldOrder {
              "direction": "DESCENDING",
              "field": FieldPath {
                "segments": Array [
                  "properties",
                ],
              },
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": Object {
            "before": false,
            "values": Array [
              Object {
                "mapValue": Object {
                  "fields": Object {
                    "info": Object {
                      "mapValue": Object {
                        "fields": Object {
                          "strength": Object {
                            "stringValue": "high",
                          },
                        },
                      },
                    },
                  },
                },
              },
            ],
          },
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with queries startAt (with integer proto field) 1`] = `
Array [
  Object {
    "data": Object {
      "ref": Query {
        "_allowUndefined": false,
        "_firestore": Object {
          "projectId": "demo",
        },
        "_queryOptions": QueryOptions {
          "allDescendants": false,
          "collectionId": "documents",
          "converter": Object {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldFilters": Array [],
          "fieldOrders": Array [
            FieldOrder {
              "direction": "ASCENDING",
              "field": FieldPath {
                "segments": Array [
                  "id",
                ],
              },
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": ResourcePath {
            "segments": Array [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": Object {
            "before": true,
            "values": Array [
              Object {
                "integerValue": 10,
              },
            ],
          },
        },
        "_serializer": Serializer {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "path": "documents/1",
    "queries": Array [
      "ref",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with timestamps 1`] = `
Array [
  Object {
    "data": Object {
      "date": Timestamp {
        "_nanoseconds": 0,
        "_seconds": 1640995200,
      },
      "id": "1",
      "nested": Object {
        "dates": Array [
          Timestamp {
            "_nanoseconds": 0,
            "_seconds": 1640995200,
          },
          Timestamp {
            "_nanoseconds": 0,
            "_seconds": 1640995200,
          },
        ],
      },
    },
    "path": "documents/1",
    "timestamps": Array [
      "date",
      "nested.dates.0",
      "nested.dates.1",
    ],
  },
]
`;

exports[`deserializeDocuments deserializes documents with timestamps 2`] = `2022-01-01T00:00:00.000Z`;

exports[`deserializeDocuments deserializes plain documents 1`] = `
Array [
  Object {
    "data": Object {
      "attachments": 2,
      "content": "Dear John,
Hello world!",
      "from": "example@email.com",
      "tags": Object {
        "important": true,
        "work": true,
      },
      "to": Array [
        "recipient1@email.com",
        "recipient2@email.com",
      ],
    },
    "path": "documents/1",
  },
]
`;
